src/ast.rs

```diff
@@ -636,13 +636,18 @@ impl Annotations {
 
 #[derive(Clone, Debug, Eq)]
 pub struct Port {
-    pub port_num: u16,
-    range: Option<Range<usize>>,
+    pub low_port_num: u16,
+    pub high_port_num: Option<u16>,
+    pub range: Option<Range<usize>>,
 }
 
 impl Port {
-    pub fn new(port_num: u16, range: Option<Range<usize>>) -> Self {
-        Port { port_num, range }
+    pub fn new(low_port_num: u16, range: Option<Range<usize>>) -> Self {
```

Comment:
This is a bit of a philosophical question, when defining an object like a port do we prefer to use new or Port {...}?  If we prefer new, I think we should probably have another new that takes in low and high.

Reply:
Broadly speaking, we prefer new(), nearly all the time.  I've been generally avoiding most proper encapsulation in the ast though, because the whole idea is for the parser to shove the code in the ast largely as it exists in source, and then the remainder of compilation is free to muck with it however it likes.  It has been my experience that adding proper encapsulation principles in that narrow use case often adds additional boilerplate and redirection without a ton of benefit, because of how transparent the AST types typically are.

That said, this might be a case where a little bit of extra encapsulation will actually be helpful, because of the combining of individual ports and port ranges into one structure, and because Port is reused as an internal structure as well in functions.rs.  That said, rust doesn't support multiple signatures for the same function, so defining an additional 3 argument new() is not possible, they two news() would need different names (eg new_range()).  Having just typed all that up, new_range() is actually a fine name.  I'm going to go ahead and add it.